pkg/gardenlet/controller/managedseed/managedseed_controller.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package managedseed
+
+import (
+	"k8s.io/client-go/tools/cache"
+
+	seedmanagementv1alpha1 "github.com/gardener/gardener/pkg/apis/seedmanagement/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils"
+)
+
+func (c *Controller) managedSeedAdd(obj interface{}) {
+	managedSeed, ok := obj.(*seedmanagementv1alpha1.ManagedSeed)
+	if !ok {
+		return
+	}
+
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+
+	if managedSeed.Generation != managedSeed.Status.ObservedGeneration {
```

Comment:
I think, we should have a shortcut for `ManagedSeeds` in deletion as well (i.e. enqueue `ManagedSeeds` with a `deletionTimestamp` without a delay).

Reply:
I don't think anything needs to be done here, it already should work as expected. When deletion timestamp becomes non-nil, generation is also increased due to https://github.com/stoyanr/gardener/blob/dea0f5e10d715214e3f33a4b19385dabc84b74bd/pkg/registry/seedmanagement/managedseed/strategy.go#L68-L71, therefore it will be enqueued immediately. This is also what I always observed during testing. Did you actually reproduce deletion being enqueued with a delay, and if yes how?
The only case I could think of that this would happen is if gardenlet restarts during the deletion. But in this case, I believe enqueuing with a jittered delay is in general what we want.

Reply:
> Did you actually reproduce deletion being enqueued with a delay, and if yes how?
The only case I could think of that this would happen is if gardenlet restarts during the deletion.

Yes, I restarted the gardenlet in order to shortcut the exponential backoff, because the `garden` namespace was hanging.

> But in this case, I believe enqueuing with a jittered delay is in general what we want.

Hmm, I don't think it would be problematic to directly enqueue ManagedSeeds in deletion on startup, as there shouldn't be that many ManagedSeeds that are marked for deletion and it doesn't cause any additional load (compared to upgrading deployed gardenlets to a newer version).

Reply:
On the other hand, for the same reasons, leaving the jittered delay in this case also doesn't hurt. Anyway, I will add the condition to he line above if you prefer it this way. There is definitely no need to change anything in line 53 below - when gardenlet restarts, all existing managedseeds are "created", therefore the `managedSeedUpdate` function is not invoked in this case. The condition there just protects from enqueuing e.g. status updates.

Reply:
Done

Reply:
> There is definitely no need to change anything in line 53 below - when gardenlet restarts, all existing managedseeds are "created", therefore the managedSeedUpdate function is not invoked in this case. The condition there just protects from enqueuing e.g. status updates.

Yes, you're right. That part is fine.
Thanks.